File: py/dqn.py
from keras import Model
from keras.optimizers import SGD, Adam
from keras.layers import Input, Conv2D, Dense, Flatten
import tensorflow as tf
import numpy as np
from collections import deque
import random

class DQNAgent:
	def __init__(self, state_size, action_size):
		# self.rows = rows
		# self.cols = cols
		self.state_size = state_size
		self.action_size = action_size
		self.memory = deque(maxlen=100000)
		self.gamma = 0.95 #discount rate
		self.epsilon = 1.0 #exploration rate
		self.epsilon_min = 0.001
		self.epsilon_decay = 0.995
		self.learning_rate = 0.001
		self.model = self.create_model()

	def create_model(self):
		inputs = Input(shape=(self.state_size,))
		hidden = Dense(128, activation='relu')(inputs)
		outputs = Dense(self.action_size, activation='softmax')(hidden)
		model = Model(inputs=inputs, outputs=outputs)

		# inputs = Input(shape=(3, self.rows, self.cols))
		# conv1 = Conv2D(16, kernel_size=(1, 1), strides=(1, 1), padding='same', activation='relu', data_format='channels_first')(inputs)
		# #conv2 = Conv2D(16, kernel_size=(3, 3), strides=(1, 1), padding='same', activation='relu', data_format='channels_first')(conv1)
		# flat = Flatten()(conv1)
		# hidden = Dense(128, activation='relu')(flat)
		# outputs = Dense(self.action_size, activation='softmax')(hidden)
		# model = Model(inputs=inputs, outputs=outputs)

		model.compile(optimizer=Adam(lr=self.learning_rate), loss='mse')
		model.summary()
		return model

	def remember(self, state, action, reward, next_state, done):
		self.memory.append((state, action, reward, next_state, done))

	def get_action(self, state):
		if np.random.rand() <= self.epsilon:
			if self.epsilon > self.epsilon_min:
				self.epsilon *= self.epsilon_decay
			idx = random.randrange(self.action_size)
			act_values = [0 for i in range(self.action_size)]
			act_values[idx] = 1
			return [act_values]
		return self.model.predict(state)

	def train_step(self, states, actions, rewards, next_states, dones):
		targets = self.model.predict(states)

		for i in range(len(dones)):
			q_new = rewards[i]
			if not dones[i]:
				next_state = np.expand_dims(next_states[i], axis=0)
				q_new = (rewards[i] + self.gamma * np.amax(self.model.predict(next_state)[0]))
			targets[i][np.argmax(actions[i])] = q_new

		history = self.model.fit(states, targets, epochs=1, verbose=0)
		loss = history.history['loss'][0]
		return loss
	
	def train_short_memory(self, state, action, reward, next_state, done):
		return self.train_step(state, action, reward, next_state, done)

	def train_long_memory(self, batch_size):
		minibatch = random.sample(self.memory, min(len(self.memory), batch_size))
		states, actions, rewards, next_states, dones = zip(*minibatch)
		states = np.squeeze(states)
		actions = np.squeeze(actions)
		rewards = np.squeeze(rewards)
		next_states = np.squeeze(next_states)
		dones = np.squeeze(dones)
		return self.train_step(states, actions, rewards, next_states, dones)

	def load(self, name):
		self.model.load_weights(name)
	
	def save(self, name):
		self.model.save_weights(name)

File: js/src/dqn.js
import * as tf from '@tensorflow/tfjs';

export default class DQNAgent {
  constructor(state_size, action_size, batch_size=32) {
    this.state_size = state_size;
    this.action_size = action_size;
    this.memory = [];
    this.memory_max_len = 10000;
    this.gamma = 0.95;
    this.epsilon = 1.0;
    this.epsilon_min = 0.001;
    this.epsilon_decay = 0.995;
    this.learning_rate = 0.001;
    this.batch_size = batch_size;
    this.model = this.create_model();
  }
	
  create_model() {
    const m = tf.sequential();
    m.add(tf.layers.dense({ inputShape: [this.state_size], units: 128, activation: 'relu' }));
    m.add(tf.layers.dense({ units: this.action_size, activation: 'softmax' }));

    m.compile({
      optimizer: tf.train.adam(),
      loss: tf.losses.meanSquaredError,
      metrics: ['mse'],
    });
    return m;
  }

  remember(state, action, reward, next_state, done) {
    this.memory.push([state, action, reward, next_state, done]);
    if (this.memory.length > this.memory_max_len) this.memory.shift();
  }

  async get_action(state) {
    // console.log('got action')
    const randVal = await tf.randomUniform([1]).arraySync()[0];
    if (randVal <= this.epsilon) {
      if (this.epsilon > this.epsilon_min) {
        this.epsilon *= this.epsilon_decay
      }
      const idx = await tf.randomUniform([1], 0, this.action_size, 'int32').arraySync()[0];
      const act_values = tf.oneHot(idx, this.action_size);
      // console.log(act_values)
      return await act_values.arraySync();
    }
    return await this.model.predict(state).arraySync()[0];
  }

  async train_step(states, actions, rewards, next_states, dones) {
    const targets = await this.model.predict(states).arraySync();

    const arrayActions = await actions.arraySync();
    const arrayDones = await dones.arraySync();
    const arrayRewards = await rewards.arraySync();
    const arrayNextStates = await next_states.arraySync();

    // console.log(arrayDones);
    for (let i = 0; i < arrayDones.length; i++) {
      // console.log('Reward: ', arrayRewards[i]);
      let q_new = arrayRewards[i];
      // console.log('Target: ', targets[i]);
      if (!arrayDones[i]) {
        const next_state = tf.expandDims(arrayNextStates[i], 0);
        const future_reward = await this.model.predict(next_state).arraySync()[0];
        // console.log('Future Reward: ', future_reward);
        // console.log('Max future reward: ', max(...future_reward));
        q_new = (arrayRewards[i] + this.gamma * Math.max(...future_reward));
        // console.log('Q new: ', q_new);
      }
      const action_idx = await tf.argMax(arrayActions[i]).dataSync()[0];
      // console.log(action_idx);
      targets[i][action_idx] = q_new;
      // console.log('Final Target: ', targets[i]);
    }

    const tensorTargets = tf.tensor(targets);

    const history = await this.model.fit(states, tensorTargets, { epochs: 1 });
    // console.log(history);
    return history.history.loss[0];
  }
	
  async train_short_memory(state, action, reward, next_state, done) {
    return await this.train_step(tf.tensor(state), tf.tensor(action), tf.tensor(reward), tf.tensor(next_state), tf.tensor(done));
  }

  async train_long_memory() {
    const minibatch = await randomSample(this.memory, Math.min(this.memory.length, this.batch_size));
    let [states, actions, rewards, next_states, dones] = zip(...minibatch);
    states = tf.tensor(states);
    actions = tf.squeeze(actions);
    rewards = tf.squeeze(rewards);
    next_states = tf.tensor(next_states);
    dones = tf.squeeze(dones);
    return await this.train_step(states, actions, rewards, next_states, dones);
  }
}

const zip = (...arr) => {
  return Array(Math.max(...arr.map(a => a.length))).fill().map((_, i) => arr.map(a => a[i]));
}

const randomSample = async (arr, sampleSize) => {
  let n = arr.length;
  const res = [];
  for (let i = 0; i < sampleSize; i++) {
    const idx = await tf.randomUniform([1], 0, n, 'int32').arraySync()[0];
    [arr[n - 1], arr[idx]] = [arr[idx], arr[n - 1]];
    n--;
    res.push(arr[idx]);
  }
  return res;
}